keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1069,187 @@ the remaining requests in that queue start getting faster service.  In
+can decide to arbitrarily adjust it in future releases.
+
+Given that the estimation is well separated piece of logic, we can decide
+to replace with much more sophisticated logic later (e.g. whether it is
+served from etcd or from cache, whether it is namespaced or not, etc.).
+
+One more important aspect to resolve is what happens if a given priority
+level doesn't have enough concurrency units assigned to it. To be on the
+safe side we should probably implement borrowing across priority levels.
+However, given we don't want to block introducing the `width` concept on
+design and implementation of borrowing, until this is done we have two
+main options:
+- cap the `width` at the concurrency units assigned to the priority level
+- reject requests for which we won't be able to allocate enough concurrency
+  units
+
+To avoid breaking some users, we will proceed with the first option.  That
+said, to accommodate for the inaccuracy here we will introduce a concept
+of `additional latency` for a request.  This basically means that after the
+request finishes in a real world, we still don't mark it as finished in
+the virtual world for `additional latency`.
+Adjusting virtual time of a queue to do that is trivial.  The other thing
+to tweak is to ensure that the concurrency units will not get available
+for other requests for that time (because currently all actions are
+triggerred by starting or finishing some request).  We will maintain that
+possibility by wrapping the handler into another one that will be sleeping
+for `additional latence` after the request is processed.
+
+Note that given the estimation for duration of processing the requests is
+automatically corrected (both up and down), there is no need to change that
+in the initial version.
+
+#### Dispatching the request
+
+The hardest part of adding support for LIST requests is dispatching the
+requests.  Now in order to start processing a request, it has to accumulate
+`<width>` units of concurrency.
+
+The important requirement to recast now is fairness.  As soon a single
+request can consume more units of concurrency, the fairness is
+no longer about the number of requests from a given queue, but rather
+about number of consumed concurrency units.  This justifes the above
+definition of adjusting the cost of the request to now be equal to
+`<width> x <processing latency>` (instead of just `<processing latency>`).
+
+At the same time, we want to maximally utilize the available capacity.
+In other words, we want to minimize the time when some concurrency unit
+is not used, but there are requests at a given PL that could use it.
+
+In order to achieve the above goals, we are introducing the following
+modification to the current dispatching algorith:
+- as soon as we dispatch the first token (concurrency unit) to a queue
+  with LIST request at its front, all further tokens are also assigned
+  to this queue until the LIST request can be started.
```

Comment:
I don't understand what this accomplishes?

Reply:
Oh, sorry, yeah now I understand.

Wouldn't it be clearer to just say that requests will be dispatched as long as their width is less than the number of seats available? (do we use the word "tokens" elsewhere for this?)

(this differs from what I'm proposing in that if you do it this way, you have to cap the max width at the number of available seats)